site/docs/guides/kubernetes.md

```diff
@@ -332,3 +332,153 @@ To uninstall the Helm chart and delete the `nessie` release from the `nessie-ns`
+JVM issues such as memory leaks, high CPU usage, etc. can be debugged using JVM tools.
+
+The Nessie container ships with few utilities, but it does have `jcmd`, a command-line utility for
+interacting with the JVM, installed.
+
+First, get a shell in the Nessie container:
+
+```bash
+kubectl exec -it -n <namespace> <pod> -c nessie -- /bin/bash
+```
+
+Then, you can use `jcmd` to capture a thread dump, heap dump, etc.:
+
+```bash
+jcmd 1 Thread.print
+jcmd 1 GC.heap_dump /tmp/heapdump.hprof
+```
+
+!!! Tip
+    Nessie server PID is usually 1. You can double-check with `ps aux` or `jps`.
+
+If you need other JVM tools, such as `jfr` or `async-profiler`, a more complex setup is required.
+
+First, restart the Nessie pod with some extra JVM options. The most
+useful option to add is the `--XX:+StartAttachListener` JVM option; without it, the JVM will not
+allow attaching to it and tools like `jstack` will fail.
+
+This can be done by modifying the pod template spec in the deployment spec, and adding/updating the
+`JAVA_OPTS_APPEND` environment variable:
+
+```bash
+java_opts_append=$(kubectl get deployment -n <namespace> <deployment> -o jsonpath='{.spec.template.spec.containers[0].env[?(@.name=="JAVA_OPTS_APPEND")].value}')
+kubectl set env -n <namespace> deployment <deployment> JAVA_OPTS_APPEND="$java_opts_append -XX:+StartAttachListener"
+```
+
+!!! Warning
+    The above command will restart all Nessie pods! Unfortunately that's inevitable, because 
+    environment variables cannot be changed in the pod spec directly.
+
+Once the target pod is ready to be attached, we can use the `lightrun-platform/koolkits/koolkit-jvm`
+image, which contains a JVM-based toolset for debugging Java applications:
+
+```bash
+kubectl debug -it -n <namespace> <pod> --image=lightruncom/koolkits:jvm --target=nessie --share-processes
+```
+
+See the [JVM KoolKits page](https://github.com/lightrun-platform/koolkits/blob/main/jvm/README.md)
+for more information. Beware that the image is quite large, so it may take some time to download.
+
+A few preliminary commands must be executed prior to be able to use the tools, because the Nessie
+process is running as a non-root user (UID 185) while the debug container is running as root (UID
+0). You can do this by creating a new user with the same UID and GID as the Nessie process, copying
+a few files and then switching to that user:
+
+```bash
+adduser --home /home/debug --uid 185 --gid 0 --disabled-password --gecos "" debug
+cp -Rf /root/.sdkman/ /home/debug/ && chown -R 185:0 /home/debug/.sdkman
+cp /root/.bashrc /home/debug/ && chown 185:0 /home/debug/.bashrc 
+su - debug
+```
+
+After running the above commands, you should be able to use `jps`, `jcmd`, `jmap`, etc. as well as
+other tools like `async-profiler`, `jfr` (Java Flight Recorder), etc. For example, you can check the
+Nessie server's JVM threads with `jcmd`, or run `async-profiler` to profile the Nessie server:
+
+```bash
+async-profiler -d 30 -f profile.html 1
```

Comment:
Sorry for dismissing your approvals, I'm trying to validate this example but async-profiler seems buggy on an arm64 machine. Switching to amd64 to see if it's better.

Reply:
Well, async-profiler in this version seems broken. I will switch the example to JFR instead which seems to be working fine.

Reply:
Ready for review again.

Reply:
Ideally we could publish our own debug image with the right user (185), all the tools (CLI, admin) and all the utilities both for network and JVM diagnostics. But that's enough for now :-)